pkg/cvo/status.go

```diff
@@ -83,10 +83,11 @@ func mergeOperatorHistory(config *configv1.ClusterVersion, desired configv1.Upda
 			}
 		}
 	} else {
-		klog.V(5).Infof("must add a new history entry completed=%t desired=%#v != last=%#v", completed, desired, last)
+		klog.V(5).Infof("must add a new history if completed or version has been set, entry completed=%t desired=%#v != last=%#v", completed, desired, last)
 		if last.CompletionTime == nil {
 			last.CompletionTime = &now
 		}
+		// add if update completed otherwise wait until version has been set to ensure correct precondition bypass check for minor version updates
```

Comment:
Are we sure the version always gets set?  When does this happen?  Is it always before the outgoing CVO pushes the updated CVO Deployment and is replaced by the incoming, new-version CVO?

Reply:
I can't say it does "always" as I do not understand the complete CVO upgrade flow "soup to nuts", e.g. "before the outgoing CVO pushes the updated CVO Deployment and is replaced by the incoming, new-version CVO".

But here's logs showing the issue:

-- here the payload is still  being loaded when version history is updated with Version=""
`I0629 19:26:24.086136       1 sync_worker.go:477] Loading payload
I0629 19:26:24.086129       1 status.go:79] merge into existing history completed=true desired=v1.Update{Version:"0.0.1-2020-06-29-183951", Image:"registry.svc.ci.openshift.org/ci-ln-dx8g8l2/release@sha256:25987f71484d28f62bd2f01246cc090c8b56250b3d73d63431ba0d24596cd08c", Force:false} last=&v1.UpdateHistory{State:"Completed", StartedTime:v1.Time{Time:time.Time{wall:0x0, ext:63729053225, loc:(*time.Location)(0x2483760)}}, CompletionTime:(*v1.Time)(0xc421810d60), Version:"0.0.1-2020-06-29-183951", Image:"registry.svc.ci.openshift.org/ci-ln-dx8g8l2/release@sha256:25987f71484d28f62bd2f01246cc090c8b56250b3d73d63431ba0d24596cd08c", Verified:false}
I0629 19:26:24.086296       1 cvo.go:441] Finished syncing cluster version "openshift-cluster-version/version" (438.176µs)
I0629 19:26:33.106047       1 payload.go:210] Loading updatepayload from "/etc/cvo/updatepayloads/ygkKOkNFlguU7-kIqUBjPQ"
I0629 19:26:33.545851       1 upgradeable.go:62] cv.Status.History[0].Version 0.0.1-2020-06-29-183951 currentMinor 0 releaseContext.DesiredVersion 4.4.0-0.nightly-2020-06-01-021027 desiredMinor 4
E0629 19:26:33.545893       1 precondition.go:59] Precondition "ClusterVersionUpgradeable" failed: Cluster operator kube-apiserver cannot be upgraded: FeatureGatesUpgradeable: "TechPreviewNoUpgrade" does not allow updates
E0629 19:26:33.545936       1 sync_worker.go:329] unable to synchronize image (waiting 2m52.525702462s): Precondition "ClusterVersionUpgradeable" failed because of "FeatureGates_RestrictedFeatureGates_TechPreviewNoUpgrade": Cluster operator kube-apiserver cannot be upgraded: FeatureGatesUpgradeable: "TechPreviewNoUpgrade" does not allow updates
I0629 19:26:36.578086       1 cvo.go:439] Started syncing cluster version "openshift-cluster-version/version" (2020-06-29 19:26:36.578069009 +0000 UTC m=+2228.491375850)
I0629 19:26:36.578264       1 cvo.go:468] Desired version from spec is v1.Update{Version:"", Image:"quay.io/openshift-release-dev/ocp-release-nightly@sha256:6d02e83e08c9ff183f9e6197117bdd3ff6598d50357b7a66eddf003b0b59f815", Force:false}
I0629 19:26:36.578315       1 status.go:159] Synchronizing errs=field.ErrorList{} status=&cvo.SyncWorkerStatus{Generation:2, Step:"PreconditionChecks", Failure:(*payload.UpdateError)(0xc421a2b720), Fraction:0, Completed:0, Reconciling:false, Initial:false, VersionHash:"", LastProgress:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}, Actual:v1.Update{Version:"", Image:"quay.io/openshift-release-dev/ocp-release-nightly@sha256:6d02e83e08c9ff183f9e6197117bdd3ff6598d50357b7a66eddf003b0b59f815", Force:false}, Verified:true}
I0629 19:26:36.578395       1 status.go:183] version=quay.io/openshift-release-dev/ocp-release-nightly@sha256:6d02e83e08c9ff183f9e6197117bdd3ff6598d50357b7a66eddf003b0b59f815
I0629 19:26:36.578489       1 status.go:51] desired.Image=quay.io/openshift-release-dev/ocp-release-nightly@sha256:6d02e83e08c9ff183f9e6197117bdd3ff6598d50357b7a66eddf003b0b59f815
I0629 19:26:36.578515       1 status.go:87] must add a new history entry completed=false desired=v1.Update{Version:"", Image:"quay.io/openshift-release-dev/ocp-release-nightly@sha256:6d02e83e08c9ff183f9e6197117bdd3ff6598d50357b7a66eddf003b0b59f815", Force:false} != last=&v1.UpdateHistory{State:"Completed", StartedTime:v1.Time{Time:time.Time{wall:0x0, ext:63729053225, loc:(*time.Location)(0x2483760)}}, CompletionTime:(*v1.Time)(0xc421682120), Version:"0.0.1-2020-06-29-183951", Image:"registry.svc.ci.openshift.org/ci-ln-dx8g8l2/release@sha256:25987f71484d28f62bd2f01246cc090c8b56250b3d73d63431ba0d24596cd08c", Verified:false}
...`
-- and a bit later in loading the version is known but the version history entry is never updated:
'...
I0629 19:29:39.024086       1 sync_worker.go:477] Loading payload
I0629 19:29:39.024255       1 payload.go:210] Loading updatepayload from "/etc/cvo/updatepayloads/ygkKOkNFlguU7-kIqUBjPQ"
I0629 19:29:39.705575       1 upgradeable.go:62] cv.Status.History[0].Version  currentMinor  releaseContext.DesiredVersion 4.4.0-0.nightly-2020-06-01-021027 desiredMinor 4
E0629 19:29:39.705611       1 precondition.go:59] Precondition "ClusterVersionUpgradeable" failed: Cluster operator kube-apiserver cannot be upgraded: FeatureGatesUpgradeable: "TechPreviewNoUpgrade" does not allow updates
E0629 19:29:39.705641       1 sync_worker.go:329] unable to synchronize image (waiting 2m52.525702462s): Precondition "ClusterVersionUpgradeable" failed because of "FeatureGates_RestrictedFeatureGates_TechPreviewNoUpgrade": Cluster operator kube-apiserver cannot be upgraded: FeatureGatesUpgradeable: "TechPreviewNoUpgrade" does not allow updates'

Reply:
I do not understand the correlation between the empty history entry and the precondition failure. The code is not  easy to understand.

Reply:
I added the following additional output right before this check:

https://github.com/openshift/cluster-version-operator/blob/8b9931ba7da67cc973bb532f42580a14c84ec0ac/pkg/payload/precondition/clusterversion/upgradeable.go#L69

Because History[0].Version="" currentMinor is also empty rather than being 4 in this case.  
 
`I0629 19:29:39.705575 1 upgradeable.go:62] cv.Status.History[0].Version currentMinor releaseContext.DesiredVersion 4.4.0-0.nightly-2020-06-01-021027 desiredMinor 4`

Reply:
> Are we sure the version always gets set? When does this happen? Is it always before the outgoing CVO pushes the updated CVO Deployment and is replaced by the incoming, new-version CVO?

In my testing it does always get set and yes, it does appear to be happening right before the current CVO shutdown (see logs below which include my additional logging) however that is only because the setting of the version allows the precondition to pass and the upgrade to proceed. WRT being sure it always gets set, I will investigate further to understand what is triggering it to be set at the later time. I'm thinking it is related to pulling in the update and getting far enough to pull the version from the pulled update. 

I0701 22:21:26.239005       1 start.go:154] Shutting down due to terminated
I0701 22:21:26.239090       1 task_graph.go:483] Canceled worker 0 while waiting for work
E0701 22:21:26.239515       1 task.go:81] error running apply for deployment "openshift-cluster-version/cluster-version-operator" (5 of 573): timed out waiting for the condition
I0701 22:21:26.239634       1 task_graph.go:483] Canceled worker 3 while waiting for work
I0701 22:21:26.239950       1 start.go:265] Stepping down as leader
I0701 22:21:26.240056       1 cvo.go:439] Started syncing cluster version "openshift-cluster-version/version" (2020-07-01 22:21:26.24004483 +0000 UTC m=+439.849473861)
I0701 22:21:26.240144       1 cvo.go:468] Desired version from spec is v1.Update{Version:"", Image:"quay.io/openshift-release-dev/ocp-release-nightly@sha256:6d02e83e08c9ff183f9e6197117bdd3ff6598d50357b7a66eddf003b0b59f815", Force:false}
I0701 22:21:26.240246       1 status.go:167] Synchronizing errs=field.ErrorList{} status=&cvo.SyncWorkerStatus{Generation:2, Step:"ApplyResources", Failure:error(nil), Fraction:0.006980803, Completed:0, Reconciling:false, Initial:false, VersionHash:"QcT0ABI6Jks=", LastProgress:time.Time{wall:0xbfb760f989960b9d, ext:439770257314, loc:(*time.Location)(0x26aaae0)}, Actual:v1.Update{Version:"4.4.0-0.nightly-2020-06-01-021027", Image:"quay.io/openshift-release-dev/ocp-release-nightly@sha256:6d02e83e08c9ff183f9e6197117bdd3ff6598d50357b7a66eddf003b0b59f815", Force:false}, Verified:true}
I0701 22:21:26.240382       1 status.go:52] return false
I0701 22:21:26.240425       1 status.go:92] must add a new history entry completed=false desired=v1.Update{Version:"4.4.0-0.nightly-2020-06-01-021027", Image:"quay.io/openshift-release-dev/ocp-release-nightly@sha256:6d02e83e08c9ff183f9e6197117bdd3ff6598d50357b7a66eddf003b0b59f815", Force:false} != last=&v1.UpdateHistory{State:"Completed", StartedTime:v1.Time{Time:time.Time{wall:0x0, ext:63729237155, loc:(*time.Location)(0x26aaae0)}}, CompletionTime:(*v1.Time)(0xc00144b120), Version:"0.0.1-2020-07-01-214723", Image:"registry.svc.ci.openshift.org/ci-ln-cxpr372/release@sha256:9c02474583c428d455d3d880b3fd1d627694047db1bc2e2708dc11a169764bc8", Verified:false}
I0701 22:21:26.240486       1 status.go:109] adding 4.4.0-0.nightly-2020-06-01-021027
I0701 22:21:26.240524       1 status.go:120] history=[]v1.UpdateHistory{v1.UpdateHistory{State:"Partial", StartedTime:v1.Time{Time:time.Time{wall:0xbfb760f98e53e26b, ext:439849807515, loc:(*time.Location)(0x26aaae0)}}, CompletionTime:(*v1.Time)(nil), Version:"4.4.0-0.nightly-2020-06-01-021027", Image:"quay.io/openshift-release-dev/ocp-release-nightly@sha256:6d02e83e08c9ff183f9e6197117bdd3ff6598d50357b7a66eddf003b0b59f815", Verified:false}, v1.UpdateHistory{State:"Completed", StartedTime:v1.Time{Time:time.Time{wall:0x0, ext:63729237155, loc:(*time.Location)(0x26aaae0)}}, CompletionTime:(*v1.Time)(0xc00144b120), Version:"0.0.1-2020-07-01-214723", Image:"registry.svc.ci.openshift.org/ci-ln-cxpr372/release@sha256:9c02474583c428d455d3d880b3fd1d627694047db1bc2e2708dc11a169764bc8", Verified:false}}
I0701 22:21:26.240786       1 task_graph.go:483] Canceled worker 10 while waiting for work
I0701 22:21:26.240840       1 task_graph.go:483] Canceled worker 15 while waiting for work
I0701 22:21:26.240885       1 task_graph.go:483] Canceled worker 7 while waiting for work
I0701 22:21:26.240927       1 task_graph.go:483] Canceled worker 14 while waiting for work
I0701 22:21:26.240990       1 task_graph.go:483] Canceled worker 12 while waiting for work
I0701 22:21:26.241049       1 task_graph.go:483] Canceled worker 13 while waiting for work
I0701 22:21:26.241106       1 task_graph.go:483] Canceled worker 6 while waiting for work
I0701 22:21:26.241150       1 task_graph.go:483] Canceled worker 4 while waiting for work
I0701 22:21:26.241194       1 task_graph.go:483] Canceled worker 9 while waiting for work
I0701 22:21:26.241235       1 task_graph.go:483] Canceled worker 8 while waiting for work
I0701 22:21:26.241304       1 task_graph.go:483] Canceled worker 5 while waiting for work
I0701 22:21:26.241315       1 task_graph.go:483] Canceled worker 2 while waiting for work
I0701 22:21:26.241328       1 task_graph.go:483] Canceled worker 1 while waiting for work
I0701 22:21:26.243355       1 task_graph.go:483] Canceled worker 11 while waiting for work
I0701 22:21:26.243464       1 task_graph.go:523] Workers finished
I0701 22:21:26.243527       1 task_graph.go:546] Result of work: [Could not update deployment "openshift-cluster-version/cluster-version-operator" (5 of 573)]
I0701 22:21:26.243595       1 sync_worker.go:817] Summarizing 1 errors
I0701 22:21:26.243684       1 sync_worker.go:821] Update error 5 of 573: UpdatePayloadFailed Could not update deployment "openshift-cluster-version/cluster-version-operator" (5 of 573) (*errors.errorString: timed out waiting for the condition)
E0701 22:21:26.246932       1 sync_worker.go:328] unable to synchronize image (waiting 21.565712806s): Could not update deployment "openshift-cluster-version/cluster-version-operator" (5 of 573)
I0701 22:21:26.348388       1 cvo.go:441] Finished syncing cluster version "openshift-cluster-version/version" (108.331351ms)
I0701 22:21:26.353675       1 cvo.go:366] Shutting down ClusterVersionOperator
I0701 22:21:26.363703       1 start.go:276] Finished shutdown

Reply:
Further analysis of this shows it to be an issue with pulling incorrect history entry as current version (see https://bugzilla.redhat.com/show_bug.cgi?id=1822513#c14).